native/android/src/main/java/com/nozbe/watermelondb/Database.kt

```diff
@@ -53,18 +53,13 @@ class Database(private val name: String, private val context: Context) {
         val rawArgs = Array(args.size, { "" })
         return db.rawQueryWithFactory(object : SQLiteDatabase.CursorFactory {
             override fun newCursor(db: SQLiteDatabase?, driver: SQLiteCursorDriver?, editTable: String?, query: SQLiteQuery): Cursor {
-                for (i in args.indices) {
-                    val arg = args[i]
-                    if (arg is String) {
-                        query.bindString(i + 1, arg)
-                    } else if (arg is Boolean) {
-                        query.bindLong(i + 1, if (arg) 1 else 0)
-                    } else if (arg is Double) {
-                        query.bindDouble(i + 1, arg)
-                    } else if (arg == null) {
-                        query.bindNull(i + 1)
-                    } else {
-                        throw (Throwable("Bad query arg type"))
+                for ((i, arg) in args.withIndex()) {
+                    when (arg) {
+                        is String -> query.bindString(i + 1, arg)
+                        is Boolean -> query.bindLong(i + 1, if (arg) 1 else 0)
+                        is Double -> query.bindDouble(i + 1, arg)
+                        null -> query.bindNull(i + 1)
+                        else -> throw (Throwable("Bad query arg type"))
                     }
                 }
                 return SQLiteCursor(driver, editTable, query)
```

Comment:
Sorry for getting in here again, but I've found more time to look at it again and I want to propose something looking even more like proper Kotlin code 

```Kotlin
        val rawArgs = Array(args.size) { "" }
        return db.rawQueryWithFactory(
            { _, driver: SQLiteCursorDriver?, editTable: String?, query: SQLiteQuery ->
                args.withIndex().forEach { (i, arg) ->
                    when (arg) {
                        is String -> query.bindString(i + 1, arg)
                        is Boolean -> query.bindLong(i + 1, if (arg) 1 else 0)
                        is Double -> query.bindDouble(i + 1, arg)
                        null -> query.bindNull(i + 1)
                        else -> throw IllegalArgumentException("Bad query arg type: ${arg::class.java.canonicalName}")
                    }
                }
                SQLiteCursor(driver, editTable, query)
            }, sql, rawArgs, null, null
        )
```
Most of that is just cosmetics, but I do think that giving name of bad type on error (and actually marking it as IllegalArgument) is going to add great deal to error readability

Reply:
applied in #1047 , thanks!